modules/imgproc/src/spatialgradient.cpp

```diff
@@ -0,0 +1,280 @@
+// or tort (including negligence or otherwise) arising in any way out of
+// the use of this software, even if advised of the possibility of such damage.
+//
+//M*/
+
+#include "precomp.hpp"
+#include "opencv2/hal/intrin.hpp"
+
+namespace cv
+{
+
+void spatialGradient( InputArray _src, OutputArray _dx, OutputArray _dy,
+                      int ksize, int borderType )
+{
+
+    // Prepare InputArray src
+    Mat src = _src.getMat();
+    CV_Assert( !src.empty() );
+    CV_Assert( src.type() == CV_8UC1 );
+    CV_Assert( borderType == BORDER_DEFAULT || borderType == BORDER_REPLICATE );
+
+    // Prepare OutputArrays dx, dy
+    _dx.create( src.size(), CV_16SC1 );
+    _dy.create( src.size(), CV_16SC1 );
+    Mat dx = _dx.getMat(),
+        dy = _dy.getMat();
+
+    // TODO: Allow for other kernel sizes
+    CV_Assert(ksize == 3);
+
+    // Get dimensions
+    const int H = src.rows,
+              W = src.cols;
+
+    // Row, column indices
+    int i = 0,
+        j = 0;
+
+    // Store pointers to rows of input/output data
+    // Padded by two rows for border handling
+    std::vector<uchar*> P_src(H+2);
+    std::vector<short*> P_dx (H+2);
+    std::vector<short*> P_dy (H+2);
+
+    int i_top    = 0,     // Case for H == 1 && W == 1 && BORDER_REPLICATE
+        i_bottom = H - 1,
+        j_offl   = 0,     // j offset from 0th   pixel to reach -1st pixel
+        j_offr   = 0;     // j offset from W-1th pixel to reach Wth  pixel
+
+    if ( borderType == BORDER_DEFAULT ) // Equiv. to BORDER_REFLECT_101
+    {
+        if ( H > 1 )
+        {
+            i_top    = 1;
+            i_bottom = H - 2;
+        }
+        if ( W > 1 )
+        {
+            j_offl = 1;
+            j_offr = -1;
+        }
+    }
+
+    P_src[0]   = src.ptr<uchar>(i_top); // Mirrored top border
+    P_src[H+1] = src.ptr<uchar>(i_bottom); // Mirrored bottom border
+
+    for ( i = 0; i < H; i++ )
+    {
+        P_src[i+1] = src.ptr<uchar>(i);
+        P_dx [i]   =  dx.ptr<short>(i);
+        P_dy [i]   =  dy.ptr<short>(i);
+    }
+
+    // Pointer to row vectors
+    uchar *p_src, *c_src, *n_src; // previous, current, next row
+    short *c_dx,  *c_dy;
+
+    int i_start = 0;
+    int j_start = 0;
+#if CV_SIMD128
+    // Characters in variable names have the following meanings:
+    // u: unsigned char
+    // s: signed int
+    //
+    // [row][column]
+    // m: offset -1
+    // n: offset  0
+    // p: offset  1
+    // Example: umn is offset -1 in row and offset 0 in column
+    uchar tmp;
+    v_uint8x16 v_um, v_un, v_up;
+    v_uint16x8 v_um1, v_um2, v_un1, v_un2, v_up1, v_up2;
+    v_int16x8 v_s1m1, v_s1m2, v_s1n1, v_s1n2, v_s1p1, v_s1p2,
+              v_s2m1, v_s2m2, v_s2n1, v_s2n2, v_s2p1, v_s2p2,
+              v_s3m1, v_s3m2, v_s3n1, v_s3n2, v_s3p1, v_s3p2,
+              v_tmp,  v_sdx, v_sdy;
+
+    for ( i = 0; i < H - 2; i += 2 )
+    {
+        p_src = P_src[i]; c_src = P_src[i+1]; n_src = P_src[i+2];
```

Comment:
It's not much overhead, but since the border handling etc is done in both the SSE and non-SSE code, this would be somewhat duplicated. Would that still be okay?

The vectors for caching pointers is only as long as the no. of rows so I thought it wouldn't have high impact.